syntax = "proto3";

import "common.proto";

package telegraph.stream;

message Subscribe {
    uint32 var_id = 1; // actually 16 bits
    uint32 min_interval = 2; // actually 16 bits
    uint32 max_interval = 3; // actually 16 bits
    uint32 timeout = 4; // actually 16 bits
}

message Call {
    uint32 action_id = 1; // actually 16 bits
    uint32 timeout = 2; // actually 16 bits
    Value arg = 3;
}

message Cancel {
    uint32 var_id = 1; // actually 16 bits
    uint32 timeout = 2; // actually 16 bits
}

message Packet {
    uint32 req_id = 1; // set to var_id for updates
    oneof event {
        Empty fetch_tree = 2;
        Node tree = 3;

        // SPECIAL: uses req_id for var_id for update
        Value update = 4;
        Call call_action = 5;
        Value call_completed = 6;
        Empty call_failed = 7;

        Subscribe change_sub = 8;
        Cancel cancel_sub = 9;
        Empty poll_sub = 10;
        bool success = 11;

        uint32 cancelled = 12;

        Empty ping = 13;
        int32 pong = 14; // contains number of subscriptions active
    }
}
